### PR TITLE
New Infrastructure Architecture

### PR DESCRIPTION
**Pull request by juanwulu**
## Description

This pull request refactors the core training and evaluation logic, as well as the model and train state interfaces, to simplify the codebase and improve clarity. The main changes include removing the old evaluation and training loop implementations, updating the `Model` and `TrainState` interfaces to be more flexible and explicit, and improving documentation throughout the code.

### Refactoring and removal of legacy code

* Removed the old evaluation loop implementation from `src/core/evaluate.py`, including all related logging and metric aggregation logic.
* Removed the old training loop implementation from `src/core/train.py`, including evaluation hooks, checkpointing, and logging.

### Model interface updates

* Refactored the `Model` interface in `src/core/model.py` to use more explicit method signatures and clearer separation between training, evaluation, and forward passes; added new methods `training_step` and `evaluation_step`, and changed `init` to return a tuple of pytrees. [[1]](diffhunk://#diff-3242de2a57605a844d12bce70704553e409b0ebf6a14497b7a822545e1535d23L29-R29) [[2]](diffhunk://#diff-3242de2a57605a844d12bce70704553e409b0ebf6a14497b7a822545e1535d23L44-R38) [[3]](diffhunk://#diff-3242de2a57605a844d12bce70704553e409b0ebf6a14497b7a822545e1535d23L53-R116) [[4]](diffhunk://#diff-3242de2a57605a844d12bce70704553e409b0ebf6a14497b7a822545e1535d23L12-R12)

### Train state improvements

* Updated `TrainState` in `src/core/train_state.py` to use standard Python dictionaries for parameters and EMA parameters, added support for mutable variables, and improved documentation.
* Enhanced the `create` method of `TrainState` to accept mutable variables and documented all arguments more clearly.

### CODE REVIEW

File: src/projects/generative/config.py [hunk 1]
```diff
@@ -150,7 +150,7 @@ def meanflow_unet_cifar_10() -> _config.ExperimentConfig:
                 path="uoft-cs/cifar10",
                 token=os.getenv("HF_TOKEN", None),
                 revision="0b2714987fa478483af9968de7c934580d0bb9a2",
-                split="test",
+                split="train",
```
[the remark this answers is not in the document]
Comment (juanwulu) [reply src/projects/generative/config.py]: This is on purpose. Will not address.

File: src/utilities/logging.py [hunk 2]
```diff
@@ -1,5 +1,69 @@
...
+            dir=work_dir,
+            group=experiment_name,
+            job_type="coordinator" if jax.process_index() == 0 else "worker",
+        )
+        _run = wandb.run
+        if not isinstance(_run, wandb_sdk.wandb_run.Run):
+            raise RuntimeError("Failed to initialize wandb run.")
+        fp = osp.join(work_dir, "wandb.txt")
+        with tf.io.gfile.GFile(fp, "w") as f:
+            f.write(_run.id)
```
[the remark this answers is not in the document]
Comment (juanwulu) [reply src/utilities/logging.py]: This works for now. Will not address.

File: src/projects/generative/experiment.py [hunk 3]
```diff
@@ -1,102 +1,87 @@
...
+            {
+                f"{prefix}/{k}": wandb.Image(np.asarray(v))
+                for k, v in outputs.images.items()
+            },
+            step=step,
+        )
+    if outputs.histograms is not None:
+        wandb.log(
+            {
+                f"{prefix}/{k}": wandb.Histogram(list(v))
```
[the remark this answers is not in the document]
Comment (juanwulu) [reply src/projects/generative/experiment.py]: This will raise `"ndarray[Any, dtype[Any]]" cannot be assigned to parameter "sequence" of type "Sequence[Unknown] | None` error.

File: src/core/model.py [hunk 4]
```diff
@@ -50,50 +44,73 @@ def init(
             **kwargs: Additional keyword arguments.
 
         Returns:
-            A `PyTree` the initialized model parameters.
+            A tuple of two pytree instances. The first contains initialized
+                model parameters and the second is
```
[the remark this answers is not in the document]
Comment (juanwulu) [reply src/core/model.py]: Fixed in d7342856e3e38b0acfdcf8eace153391124e12d5

File: src/projects/generative/experiment.py [hunk 5]
```diff
@@ -1,102 +1,87 @@
...
-    params: PyTree,
     batch: typing.Dict[str, typing.Any],
     fid_metric: fid.FrechetInceptionDistance,
     **kwargs,
 ) -> _model.StepOutputs:
     r"""Conduct a single evaluation step and compute metrics.
 
     Args:
+        params (PyTree): The model parameters.
         rngs (jax.Array): Random key for sampling.
```
[the remark this answers is not in the document]
Comment (juanwulu) [reply src/projects/generative/experiment.py]: Not a problem.

File: src/projects/rl/main.py [hunk 6]
```diff
@@ -265,10 +234,9 @@ def main(_: typing.List[str]) -> int:
                     # Forward pass (Note: pure exploitation)
                     # Add batch dimension [None, :] because the model expects
                     #  (batch, features)
+                    state = training.shard(state[None, :])
                     q_values = p_eval_step(
```
[the remark this answers is not in the document]
Comment (juanwulu) [reply src/projects/rl/main.py]: Will address in the next minor version.
